user_settings: Add option to not mark messages as read on scroll.

Joelute opened the pull request:
These changes introduces an new user option where it allows users read messages and scroll through their feed without marking those messages as read. 

Added a new unread banner that displays whenever the user has unread messages in the message feed and the `do not mark as read setting` on.

Fixes #22804.

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://user-images.githubusercontent.com/62449508/223600391-1afd345f-c2c3-4c2b-bfb6-6c61335b65fd.png)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Joelute commented:
Hello @alya! I think this PR is ready for review, please have a look!

sahil839 commented on zerver/migrations/0426_realmuserdefault_do_not_mark_messages_as_read_and_more.py:
```diff
@@ -0,0 +1,22 @@
+# Generated by Django 4.1.6 on 2023-02-06 02:01
```
The name for migration file could be better here. Something like `0426_add_do_not_mark_messages_as_read_setting`

sahil839 reviewed (commented):
Posted a couple of comments. Looks good overall.

One doubt I am having is what do when the user marked some messages as unread using the message popover and then changed the setting to mark messages read as scroll. Do we want the messages to be marked as read on scrolling in that case. I don't think this will occur frequently but just thought of mentioning it once.
![mark-read-on-scroll](https://user-images.githubusercontent.com/35494118/219107243-84008fc9-353a-47a7-8e10-47600c18ce90.gif)

Joelute commented:
Thank you for the review @sahil839! I'll work on adding the changes later today when I have time to. 

I understand the concerns of the case you described and I think we should bring this up for discussion in CZO #design. 
When I was designing and testing this, if you had the setting set to do not mark messages as read, the compose banner should continue to display. Hence, I decided to hide the banner and resume reading when the user set the setting to mark messages as read. I also thought that when a user decides to turn the setting to mark messages as read, they understand that messages will begin to be marked as read and they're fully commiting to it.

Joelute commented on zerver/openapi/zulip.yaml:
```diff
@@ -8825,6 +8825,14 @@ paths:
+            feed.
+          schema:
+            type: boolean
+          example: false
```
I've addressed these suggestions. Please have a look! :)

sahil839 commented:
Changes look good. You also need to bump `API_FEATURE_LEVEL` in `version.py`.

Yes, starting the discussion in #design sounds good. Just to clarify, I do not think current behavior is bug, but just wanted to make sure we can hve feedback of others on this behavior.

Joelute commented:
Whoops forgot about that, and done!

Sounds good! I'll start it close to next week when Alya and Tim comes back.

Joelute commented:
Hello @alya! May I get a review on this PR? I've pushed new changes that addresses issue we talked about in CZO. Also, the 4 failing tests are production suites and it's happening due to an outage.

Joelute commented:
Resolved conflict, awaiting for review.

alya commented:
Thanks! Playing with the PR, I think we are going to need a custom banner after all, as otherwise the explanation of why messages aren't being marked as read seems confusing. Can we replace the current banner with the following?

> Your Zulip app is [configured](/help/marking-messages-as-read) not to mark messages as read on scroll. [Mark as read].

The logic for when to show it can be just when you have unreads in the current view (same as with the current banner). And we can just never show the other banner if we are showing this one. E.g., just this one would be shown if Zulip is configured not to mark messages as read *and* you just marked some messages as unread in your current view.

Joelute commented:
@alya I noticed that you made "configured" into a link in the example. Should I also do that and link it to the doc instead or was that by accdident?

sahil839 commented:
> @alya I noticed that you made "configured" into a link in the example. Should I also do that and link it to the doc instead or was that by accdident?

Yes, I think you have to add that link to the doc.

alya commented:
Yes, of course, that's intentional.

Joelute commented:
Alright sounds good. I'm currently stuck on creating a new compose banner and waiting for help in [CZO](https://chat.zulip.org/#narrow/stream/49-development-help/topic/New.20compose.20banner)

Joelute commented:
Hello @alya! I've added the new custom unread banner, please have a look.

alya commented:
Nice, looks good to me in manual testing!

@amanagr Could you please review this PR?

Joelute commented:
Found edge cases where the new custom unread banner would display instead of the default preserved reading state banner on searches and near tags. Updated the commit to fix that issue. 

I also changed some function names to more generic names for better readibility.

Joelute commented:
> The logic for when to show it can be just when you have unreads in the current view (same as with the current banner). And we can just never show the other banner if we are showing this one. E.g., just this one would be shown if Zulip is configured not to mark messages as read and you just marked some messages as unread in your current view.

Hi @alya, I'm so sorry, I didn't catch that part. Currently the logic is to continue display the preserve reading state banner just like how it is currently (E.g. marking messages as unread, in a near message view, in search view), but just replacing the new custom banner if it is displayed. I'm wondering if that's still okay or not. I'm thinking that if the current logic isn't clean on the technical side, I could always go back to just simply displaying one of the two banners at all times.

amanagr commented:
LGTM

alya commented:
> Currently the logic is to continue display the preserve reading state banner just like how it is currently (E.g. marking messages as unread, in a near message view, in search view), but just replacing the new custom banner if it is displayed.

Ah, hm, I did test that we never get the two banners at the same time, but forgot about this detail of the spec too. It's probably worth raising the question in #design for feedback.

Joelute commented:
Opened discussion in [CZO #design ](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20scroll.20down.20and.20unread).

Joelute commented:
Hello @alya! This PR is ready for another manual review :)

alya commented:
Please rebase this PR and renumber the migrations. I'm seeing:

```
Conflicting migrations detected; multiple leaf nodes in the migration graph: (0433_add_do_not_mark_messages_as_read_setting, 0434_create_nobody_system_group in zerver).
To fix them run 'python manage.py makemigrations --merge'

Subcommand of /srv/zulip/tools/lib/provision_inner.py failed with exit status 1: env DJANGO_SETTINGS_MODULE=zproject.settings ZULIP_DB_NAME=zulip ./manage.py migrate --no-input
Actual error output for the subcommand is just above this.
```

Joelute commented:
Whoops, missed that in the last rebase. Working on it!

alya commented:
@Joelute Would you like to take a pass at adding desktop/web instructions to http://zulip.com/help/marking-messages-as-read#configure-whether-messages-are-automatically-marked-as-read. We'll need to do that either in this PR or as an immediate follow-up, as otherwise the link from the banner to the help page will be confusing.

Joelute commented:
Sure! I planned on doing it as a follow up after this PR.

timabbott commented on zerver/models.py:
```diff
@@ -1509,6 +1509,7 @@ class UserBaseSettings(models.Model):
     # This setting controls which view is rendered first when Zulip loads.
     # Values for it are URL suffix after `#`.
     default_view = models.TextField(default="recent_topics")
+    do_not_mark_messages_as_read = models.BooleanField(default=False)
```
This name sounds like we'll never mark messages as read, which is confusing. We should give especially careful thought to the names of settings like this that will be part of the API.

Can you start a thread in "#api design" about this value? I think we'll want to discuss both what to call it, but also whether we want a slightly different format (the mobile and web apps may want the same or different configuration here, and I'm just not sure how best to manage that in terms of what this is called).

One idea would be something like `web_app_mark_read_on_scroll`, if we want to make this clearly explicitly not controlling the mobile app behavior.

timabbott commented on web/src/unread_ui.js:
```diff
@@ -13,20 +13,45 @@ import * as topic_list from "./topic_list";
+export function update_unread_banner() {
+    $("#mark_as_read_turned_off_banner").html(render_mark_as_read_turned_off_banner());
+
+    $("#mark_view_read").on("click", () => {
```
Calling this more than once is dangerous -- one can end up with multiple copies of the click handler running on a single click. We should leave this in `initialize`, possibly with a `$("body").on("click", "#mark_view_read"` structure.

timabbott replied:
Yeah, you'll notice we use that style of click handler hanging off body a lot in the codebase, because it's more robust in this way.

timabbott commented on web/src/stream_events.js:
```diff
@@ -173,7 +173,7 @@ export function mark_unsubscribed(sub) {
         // This update would likely be better implemented by having it
         // disappear whenever no unread messages remain.
-        unread_ui.hide_mark_as_read_turned_off_banner();
+        unread_ui.hide_unread_banner();
```
Can you do this renaming as a prep commit? It'll make the actual functional changes way easier to read.

timabbott replied:
I'm also not sure the rest of this commit beyond the renaming makes sense as separated from the final commit, I guess it depends on how you resolve the other feedback.

Joelute replied:
Yeah, after making a separate commit for renaming, it doesn't actually make sense for that additional commit to be separated from the final commit. I'll have it squashed with the final commit.

timabbott commented on web/templates/mark_as_read_disabled_banner.hbs:
```diff
@@ -0,0 +1,12 @@
+    {{/tr}}
+</p>
+<div id="mark_as_read_controls">
+    <button id="mark_view_read" class="btn btn-warning" title="{{t 'Mark as read' }}">
```
Should this `title` be here? It seems redundant and not helpful -- you generally only need titles for buttons that don't have a normal label.

Joelute replied:
Ahh, that makes sense. I've removed it from both unread banners.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -675,6 +677,19 @@ export function dispatch_normal_event(event) {
+                ) {
+                    message_lists.current.resume_reading();
+                    unread_ui.hide_unread_banner();
+                }
```
This is a lot more business logic than we usually like to put directly in `server_events_dispatch`; can `unread_ui.update_unread_banner` do this?

Joelute replied:
That's a good idea, since the logic between the two was similar I was able to shorten the code down and removed all this logic here.

timabbott commented on web/src/settings_config.ts:
```diff
@@ -39,6 +39,23 @@ export const demote_inactive_streams_values = {
+        description: $t({defaultMessage: "Always"}),
+    },
+    // The `conversation_only` is still in development. Looking to
+    // integration `always` and `never` options first.
```
This plan is fine, but the second sentence can be dropped here :).

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -8994,6 +8994,20 @@ paths:
+            **Changes**: New in Zulip 7.0 (feature level 173).
+          schema:
+            type: integer
+            enum:
+              - 1
+              - 2
+              - 3
```
The description should document what these numeric values mean. For an example, see `color_scheme` just below.

Joelute replied:
Sounds good, will make the changes later.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -11037,6 +11051,13 @@ paths:
                             description: |
                               This setting has no effect at present. It is reserved for use in controlling
                               the default font size in Zulip.
+                          web_mark_read_on_scroll_policy:
```
(And similarly in the other places this property appears, like this one.)

timabbott commented on web/src/settings_config.ts:
```diff
@@ -39,6 +39,22 @@ export const demote_inactive_streams_values = {
+    // The `conversation_only` is still in development.
+    // conversation_only: {
+    //    code: 2,
+    //    description: $t({defaultMessage: "Only in conversation view"}),
```
`Only in conversation view` is probably the right spelling for this.

timabbott commented on web/src/unread_ui.js:
```diff
@@ -114,8 +131,9 @@ export function initialize() {
         hide_unread_banner();
         user_closed_unread_banner = true;
     });
+    hide_unread_banner();
```
I don't understand why you're calling both `update_unread_banner` and `hide_unread_banner` during the initialization process here. Was this a workaround for the previous selector approach?

Joelute replied:
`update_unread_banner` only updates the banner that should be shown, it doesn't exactly update the visibility of the banners. During the initalization process, the banners will be visible off start and we probably want to hide it.

timabbott replied:
OK I'm going to do this tweak to make this somewhat strange sequence of steps easy to understand:

```
diff --git a/web/src/unread_ui.js b/web/src/unread_ui.js
index d2a3a6cdbf..819d8636a7 100644
--- a/web/src/unread_ui.js
+++ b/web/src/unread_ui.js
@@ -115,7 +115,6 @@ export function should_display_bankruptcy_banner() {
 
 export function initialize() {
     update_unread_counts();
-    update_unread_banner();
     $("body").on("click", "#mark_view_read", () => {
         // Mark all messages in the current view as read.
         //
@@ -135,5 +134,11 @@ export function initialize() {
         hide_unread_banner();
         user_closed_unread_banner = true;
     });
+
+    // The combination of these functions in sequence ensures we have
+    // at least one copy of the unread banner in the DOM, invisible;
+    // this somewhat strange pattern allows our CSS to reserve space for
+    // the banner, to avoid scroll position jumps when it is shown/hidden.
+    update_unread_banner();
     hide_unread_banner();
 }
```

timabbott commented:
I expanded all of the "Changes" entries with a "Previously" sentence -- that's a habit we try to use for all API changes entries, even though in this case it's probably not very important, because the setting is not likely to be used by other clients.

With the tweaks noted above, this looks finished to me, merged, thanks @Joelute! @alya @drrosa there's one release-critical follow-up, which is to document the setting on /help/mark-messages-as-read, and possibly any other places. 

(It's likely @Joelute will manage to add the "only in conversation views" variant of this setting in the next week, since I expect it's a fairly small chunk of work, so we can potentially draft that documentation under the assumption that we're going to do that).

timabbott commented:
Oh, I see we already have https://github.com/zulip/zulip/pull/24942 open for the documentation change.

timabbott commented:
One more note -- @Joelute I caught that one of the two /help/ links was broken (just a typo -- `/help/mark-` rather than `/help/marking-`. It's a very good habit to always check links actually work in anything you're changing; my experience is the change that something I do with a link that I don't test to fail is like 25% :)

Joelute commented:
Oopies, I thought I had tested it before but I guess not.
